cmd/juju/common.go

```diff
@@ -136,9 +157,105 @@ func resolveCharmURL(curlStr string, csParams charmrepo.NewCharmStoreParams, rep
 	return curl, repo, nil
 }
 
-// charmStoreParams is called to obtain the parameters for connecting
-// to the charm store. It is defined as a variable so it can be changed
-// for testing purposes.
-var charmStoreParams = func() (charmrepo.NewCharmStoreParams, error) {
-	return charmrepo.NewCharmStoreParams{}, nil
+// addCharmViaAPI calls the appropriate client API calls to add the
+// given charm URL to state. For non-public charm URLs, this function also
+// handles the macaroon authorization process using the given csClient.
+// The resulting charm URL of the added charm is displayed on stdout.
+func addCharmViaAPI(client *api.Client, ctx *cmd.Context, curl *charm.URL, repo charmrepo.Interface, csclient *csClient) (*charm.URL, error) {
+	switch curl.Schema {
+	case "local":
+		ch, err := repo.Get(curl)
+		if err != nil {
+			return nil, err
+		}
+		stateCurl, err := client.AddLocalCharm(curl, ch)
+		if err != nil {
+			return nil, err
+		}
+		curl = stateCurl
+	case "cs":
+		if err := client.AddCharm(curl); err != nil {
+			if !params.IsCodeUnauthorized(err) {
+				return nil, errors.Mask(err)
+			}
+			m, err := csclient.authorize(curl)
+			if err != nil {
+				return nil, errors.Mask(err)
+			}
+			if err := client.AddCharmWithAuthorization(curl, m); err != nil {
+				return nil, errors.Mask(err)
+			}
+		}
+	default:
+		return nil, fmt.Errorf("unsupported charm URL schema: %q", curl.Schema)
+	}
+	ctx.Infof("Added charm %q to the environment.", curl)
+	return curl, nil
+}
+
+// newCharmStoreClient is called to obtain a charm store client
+// including the parameters for connecting to the charm store, and
+// helpers to save the local authorization cookies and to authorize
+// non-public charm deployments. It is defined as a variable so it can
+// be changed for testing purposes.
+var newCharmStoreClient = func() (*csClient, error) {
+	jar, err := cookiejar.New(&cookiejar.Options{
+		PublicSuffixList: publicsuffix.List,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	file := path.Join(utils.Home(), ".go-cookies")
```

Comment:
consider

    if file := os.Getenv("JUJU_COOKIEFILE"); file != "" {
         return file
    }

It is used elsewhere.